### ISSUE TITLE
Add an upspin command to check the validity of a file/directory.

### ISSUE THREAD

**Issue by foolusion** [upspin/upspin#460]
There should be an upspin command to check that a file stored in your upspin server matches the local file. It should report whether the file exists in upspin, the files blocks have matching checksums and all blocks are uploaded.

If you pass a local directory and an upspin directory it should check all files in the folder and produce a list of files that are different, and files missing from upspin. I'm not sure if files missing locally would be a concern, but I could imagine someone has that use case. There could also be a flag to recur through sub-directories.

**Comment by foolusion** [comment]
Previous discussion https://groups.google.com/forum/#!topic/upspin/UM2wOsA7Occ @adg @presotto @robpike 

I'm still pretty new to this project. I assume that if you got a DirEntry from your dir server all the references would exist in your store server. I also assume that data in the store server is at least equivalent to it's checksum (key).

I don't want to download any the files in order to validate if it's possible. In my case, I have hundreds of GB of photos, and I want to reclaim the space on my local HD. Would checking the upspin DirEntry references exist in the store server, and then packing the local file with the same keys, to create the same blocks/checksum, be sufficient? Also, to @adg's point should there be a separte object/function to pack data with a certain key (the one used to originally pack the file)?

**Comment by robpike** [comment]
For your situation, it may be fine to assume all the references are intact, but others may not. This suggests a flag might be needed on the tool to enable more thorough checking.

There is no Upspin mechanism to test if a reference is present without reading it. That might change to improve this tool although I don't want to immediately go to that step.

**Comment by foolusion** [comment]
https://upspin-review.googlesource.com/#/c/upspin/+/13020 cmd/upspin: add a validate sub command.
I've uploaded my current work

**Comment by robpike** [comment]
Having seen the code, I now believe it's extremely difficult to do this correctly in any mechanism other than to read both the local and Upspin files. Not impossible, but very difficult and I believe not worth doing.

Maybe it's possible to block the local file in the same blocking as the Upspin one and compare hashes, to avoid reading the remote blocks. But I also think that not reading the remote blocks is a mistake, or at least not providing the ability to do so is a mistake.

At this point I am inclined to say "no" to this.

**Comment by foolusion** [comment]
I agree. With an added struct for repacking blocks using a certain key it might be cleaner, but that only verifies the DirEntry and not the actual data. We still need a way to check that all blocks are uploaded (this might only be an issue with cacheserver). I'm fine closing this for now.

**Comment by robpike** [comment]
Leaving open because the problem is real, but I will drop the CL. (It can be recovered later if something develops.)
